### PR TITLE
Customizing host path for dynamically provisioned PersistentVolumes

### PR DESCRIPTION
**Pull request by nanikjava**
#5144 

This fix contains few things:

* Used k8s.gcr.io/debian-base-amd64:v2.0.0 as base image to build storage-provisioner
* Modify RBAC permission used to cluster-admin

### CODE REVIEW

File: deploy/addons/storage-provisioner/storage-provisioner.yaml.tmpl [hunk 1]
```diff
@@ -31,7 +31,7 @@ metadata:
 roleRef:
   apiGroup: rbac.authorization.k8s.io
   kind: ClusterRole
-  name: system:persistent-volume-provisioner
+  name: cluster-admin
```
Comment (nanikjava) [inline deploy/addons/storage-provisioner/storage-provisioner.yaml.tmpl]: This permission works fine but uncertain whether this is the correct one to use. Tried using the following

system:controller:expand-controller
system:controller:persistent-volume-binder

but it always complain permission issue either with persistentvolume/persistentvolumeclaims.

Tried defining customised RBAC inside storage-provisioner.yaml.tmp also does not work as it keep on thrown permission issue with persistentvolume(claims) and endpoints (using ClusterRole similar to defined here https://github.com/pragkent/aliyun-disk-provisioner/blob/38680a9607b0d33567be1bf0a7c57b26f0960549/deploy/rbac.yaml)

File: cmd/storage-provisioner/main.go [hunk 2]
```diff
@@ -25,15 +25,18 @@ import (
 	"k8s.io/minikube/pkg/storage"
 )
 
+var pvDir = "/tmp/hostpath-provisioner"
+
 func main() {
 	// Glog requires that /tmp exists.
+	fmt.Println("This is changes is called")
```
Comment (Eelis) [inline cmd/storage-provisioner/main.go]: Is this intentional? :)

Comment (nanikjava) [reply cmd/storage-provisioner/main.go]: > Is this intentional? :)

Good catch. Will fix it. Thanks.

File: deploy/storage-provisioner/Dockerfile [hunk 3]
```diff
@@ -12,6 +12,6 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-FROM scratch
-COPY out/storage-provisioner storage-provisioner
-CMD ["/storage-provisioner"]
+FROM k8s.gcr.io/debian-base-amd64:v2.0.0
```
Comment (tstromberg) [inline deploy/storage-provisioner/Dockerfile]: Can this stay `scratch`?

File: cmd/storage-provisioner/main.go [hunk 4]
```diff
@@ -25,6 +25,8 @@ import (
 	"k8s.io/minikube/pkg/storage"
 )
 
+var pvDir = "/tmp/hostpath-provisioner"
```
Comment (tstromberg) [inline cmd/storage-provisioner/main.go]: This location feels weird. Is this a placeholder?

Comment (afbjorklund) [reply cmd/storage-provisioner/main.go]: It's the paths that we are currently using, it gets bind-mounted/symlinked onto the persistent disk.

See https://minikube.sigs.k8s.io/docs/reference/persistent_volumes/  (it also uses `/tmp/hostpath_pv`)

File: deploy/storage-provisioner/Dockerfile (hunk 3 above)
Comment (afbjorklund) [reply deploy/storage-provisioner/Dockerfile]: I think we need to provide a build container for the copied file, right now it cheerfully builds it on the host ?

Use the normal MINIKUBE_BUILD_IN_DOCKER shenaningans, I don't see what it has do with path though...

Comment (nanikjava) [reply deploy/storage-provisioner/Dockerfile]: > Can this stay `scratch`?

Can I resolve this conversation ? as it's clear now in the slack channel conversation that we cannot use scratch ?

Comment (afbjorklund) [reply deploy/storage-provisioner/Dockerfile]: As long as you build static binaries (using CGO_ENABLED=0), then `scratch` should be fine.

What issues are you seeing ? We know that a _dynamic_ binary doesn't work with scratch...
But that was probably not the intention, but just a side-effect when redoing the make earlier.

I added the build container in #6257 if you want to have a look.

**Comment by afbjorklund** [comment]
Is it possible to merge this PR with the previous one #5400 - with the same content ?

**Comment by tstromberg** [comment]
Looks good, do you mind addressing the merge conflict?

**Comment by GrahamDumpleton** [comment]
Can someone explain to me the justification for having the file system directory created use the PVC name instead of the PV name. In other words, the change at line 60 from:

```
path := path.Join(p.pvDir, options.PVName)
```

to:

```
path := path.Join(p.pvDir, options.PVC.Name)
```

This means that two different applications in different namespaces can't use the same persistent volume claim name because if they do, they end up using the same directory on the filesystem, and thus they corrupt each other if directories/file names they use overlap inside the volume. This can break so many applications. It basically means you can't deploy multiple instances of the one application in different namespaces.

This introduces a serious bug and is not a feature.

**Comment by Eelis** [comment]
> This introduces a serious bug

Ah good point, I hadn't considered this.

> The obvious solution to avoid for this would have been to use path of form `root/namespace/pvc-name` if wanted more user friendly name.

Sounds good to me!

**Pull request merged**